### PR TITLE
Match output of 'git describe --first-parent'

### LINKED ISSUE #46: Change plugin to match behavior of "git describe --first-parent"
Currently, the output of `gradle-git-version` matches the output of `git describe`. Although this is fine in the common case, it's often not the desired behavior if merge commits exist.

For example, consider the following case:

* Mainline branch has some tag (`1.0.0-dev`) and development is proceeding normally
* Some other branch (for example, `1.0.0-hotfix`) is merged into the mainline branch
* `gradle-git-version` is run on the mainline branch

In this scenario, we typically want the output to still be `1.0.0-dev-*`. However, because `git describe` returns the nearest ancestor tag, the merge causes `1.0.0-hotfix` to be found as a parent tag that is "closer" to the commit than `1.0.0-dev`, so you get `1.0.0-hotfix-*` instead.

The fix would be to match the behavior of `git describe --first-parent` instead -- this command only considers the first parent commit for merge commits, and would cause this to work correctly. I believe that this is almost always the desired behavior.

### PR DESCRIPTION
**Pull request by nmiyake**
Makes it so that, if a merge causes a tag to be added to the
history, the tags from the merge will not be used as part of
the output for the description.

Fixes #46

### CODE REVIEW

**Comment by nmiyake** [comment]
@uschi2000 @markelliot

**Comment by nmiyake** [comment]
Ping for feedback

File: src/main/groovy/com/palantir/gradle/gitversion/DescribeFirstParentCommand.java [hunk 1]
```diff
@@ -0,0 +1,304 @@
+package com.palantir.gradle.gitversion;
```
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/gitversion/DescribeFirstParentCommand.java:61]: needs copyright and link to origin

File: src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy [hunk 2]
```diff
@@ -44,7 +45,10 @@ class GitVersionPlugin implements Plugin<Project> {
     private String gitDesc(Project project) {
         Git git = gitRepo(project)
         try {
-            String version = git.describe().call() ?: UNSPECIFIED_VERSION
+            DescribeFirstParentCommand describe = new DescribeFirstParentCommand(git.getRepository())
+//            DescribeCommand describe = new DescribeCommand(git.getRepository())
```
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy]: huh?

Comment (nmiyake) [reply src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy]: Oops sorry poor form on my part -- debugging code that should be removed.

File: src/main/groovy/com/palantir/gradle/gitversion/DescribeFirstParentCommand.java (hunk 1 above)
Comment (nmiyake) [reply src/main/groovy/com/palantir/gradle/gitversion/DescribeFirstParentCommand.java:61]: Yup had this on my todo list but forgot to do before push -- will do

**Review by uschi2000 — APPROVED** [review]
i'm fine with this, but sort of think that relying on OS `git` is simpler?

**Comment by nmiyake** [comment]
Yup, just a matter of trade-offs. If we depend on OS git, we'll need to assume that the host system has `git` installed, that Gradle can call it using `git` and that it will resolve to the correct command, and that all of the versions of `git` we expect to use are of a version that has the commands/flags that we expect. It will also be a bigger code change for this Gradle plugin.

I don't have a strong preference either way. Even though this change does add a bit of code, I think that the surface area is small (and targeted) enough that I think it makes sense. We can also still migrate to shelling out later if we decide that we want to do that (or find more Git features that we want to use).

If you would prefer that we use shell out to Git, happy to open a PR that converts this plugin to do that. However, for now I would lean towards adding this functionality in this manner in JGit first.

**Comment by nmiyake** [comment]
Addressed PR feedback

**Comment by uschi2000** [comment]
+1

**Comment by nmiyake** [comment]
Thanks! @markelliot are you good with this change?

File: src/test/groovy/com/palantir/gradle/gitversion/GitVersionPluginTests.groovy [hunk 3]
```diff
@@ -155,6 +156,40 @@ class GitVersionPluginTests extends Specification {
...
+        // switch back to main branch and merge hotfix branch into main branch
+        git.checkout().setName(master).call()
+        git.merge().include( git.getRepository().getRef("hotfix")).setFastForward(MergeCommand.FastForwardMode.NO_FF).setMessage("merge commit").call()
+
+        when:
+        BuildResult buildResult = with('printVersion').build()
+
+        then:
+        buildResult.output =~ ":printVersion\n1.0.0-1-g[a-z0-9]{7}\n"
+    }
```
Comment (markelliot) [inline src/test/groovy/com/palantir/gradle/gitversion/GitVersionPluginTests.groovy:193]: I'd like to see the same test but also tagging a 2.0.0 after the hotfix merges to ensure we see that come up.

Comment (nmiyake) [reply src/test/groovy/com/palantir/gradle/gitversion/GitVersionPluginTests.groovy:193]: Done

**Comment by nmiyake** [comment]
Addressed PR feedback

**Review by markelliot — APPROVED** [review]

**Comment by nmiyake** [comment]
@markelliot @uschi2000 can we tag a 0.5.4 or 0.6.0 to publish a release that includes this change?

**Comment by uschi2000** [comment]
I don't have permissions, sorry.

**Comment by markelliot** [comment]
@uschi2000 you should now

**Comment by nmiyake** [comment]
Ping

**Comment by uschi2000** [comment]
released as 0.6.0

**Comment by nmiyake** [comment]
Thanks!

**Pull request merged**